tests/integration/test_no_questionnaire_state.py

```diff
@@ -0,0 +1,41 @@
+from mock import patch
+
+from tests.integration.integration_test_case import IntegrationTestCase
+
+
+class TestNoQuestionnaireState(IntegrationTestCase):
+    def test_no_state_before_request(self):
+        # Given
+        self.launchSurvey("test_view_submitted_response")
+
+        # When
+        with patch("app.routes.questionnaire.get_metadata", return_value=None):
+            self.post()
+
+            # Then
+            self.assertStatusUnauthorised()
+
+    def test_questionnaire_not_submitted_no_state_before_submission_request(self):
+        # Given
+        self.launchSurvey("test_view_submitted_response")
+
+        # When
+        with patch("app.routes.questionnaire.get_metadata", return_value=None):
+            self.get("/submitted/view-response")
+
+            # Then
+            self.assertStatusUnauthorised()
+
+    def test_questionnaire_submitted_no_state_before_submission_request(self):
```

Comment:
I think the test names are confusing, how about:
- `test_questionnaire_route_before_questionnaire_submitted`
- `test_post_submission_route_before_questionnaire_submitted`
- `test_post_submission_route_after_questionnaire_submitted`

It also highlights that there should be a `test_questionnaire_route_after_questionnaire_submitted` test.

Reply:
test_questionnaire_route_after_questionnaire_submitted is an interesting one as looking at it I think there is a bug in the current code, not what has been added ```@questionnaire_blueprint.before_request
@login_required
@with_questionnaire_store
def before_questionnaire_request(questionnaire_store):
    if questionnaire_store.submitted_at:
        return redirect(url_for("post_submission.get_thank_you"))

    if request.method == "OPTIONS":
        return None

    metadata = get_metadata(current_user)
    if not metadata:
        raise NoQuestionnaireStateException(401)```

Surely the metadata /NoQuestionnaireStateException should above questionnaire_store.submitted_at

Reply:
Will fix and add the test

Reply:
Yes, that needs to be moved, it should be:
- OPTIONS check
- questionnaire state check
- submitted at check